pkg/messaging/mqtt/pubsub.go

```diff
@@ -73,14 +72,28 @@ func (ps pubsub) Subscribe(id, topic string, handler messaging.MessageHandler) e
 		return errEmptyTopic
 	}
 	ps.mu.Lock()
-	defer ps.mu.Unlock()
 	// Check client ID
 	s, ok := ps.subscriptions[id]
 	switch ok {
 	case true:
 		// Check topic
 		if ok = s.contains(topic); ok {
-			return errAlreadySubscribed
+			// Unlocking, so that Unsubscribe() can access ps.subscriptions
+			ps.mu.Unlock()
+			if err := ps.Unsubscribe(id, topic); err != nil {
+				return err
+			}
+			ps.mu.Lock()
+			if len(ps.subscriptions) == 0 {
+				client, err := newClient(ps.address, id, ps.timeout)
+				if err != nil {
+					return err
+				}
```

Comment:
What if `newClient` call fails? You locked the mutex and exit - it's a deadlock.